### PR TITLE
 update retrying when api call throttled 

### PR DESCRIPTION
**Pull request by outscale-toa**
#106

Signed-off-by: Thiery Ouattara <opensource@outscale.com>

### CODE REVIEW

File: outscale/resource_outscale_volume_test.go [hunk 1]
```diff
@@ -112,7 +112,7 @@ func testAccCheckOAPIVolumeExists(n string, v *oscgo.Volume) resource.TestCheckF
 		err = resource.Retry(5*time.Minute, func() *resource.RetryError {
 			response, _, err = conn.VolumeApi.ReadVolumes(context.Background()).ReadVolumesRequest(request).Execute()
 			if err != nil {
-				if strings.Contains(err.Error(), "RequestLimitExceeded:") {
+				if strings.Contains(err.Error(), "Request rate exceeded") {
 					return resource.RetryableError(err)
```
Comment (outscale-mdr) [inline outscale/resource_outscale_volume_test.go:116]: Why don't you use the refactored functions ?

File: outscale/resource_outscale_volume.go [hunk 2]
```diff
@@ -245,18 +238,15 @@ func resourceOAPIVolumeDelete(d *schema.ResourceData, meta interface{}) error {
 			VolumeId: d.Id(),
 		}
 		_, _, err := conn.VolumeApi.DeleteVolume(context.Background()).DeleteVolumeRequest(request).Execute()
-		if err == nil {
+		if err != nil {
 			return nil
+			if strings.Contains(fmt.Sprint(err), "VolumeInUse") {
+				return resource.RetryableError(fmt.Errorf("Outscale VolumeInUse - trying again while it detaches"))
+			}
+			return utils.CheckThrolling(err)
```
Comment (outscale-mdr) [inline outscale/resource_outscale_volume.go:246]: LGTM warning :)

File: outscale/resource_outscale_virtual_gateway_test.go [hunk 3]
```diff
@@ -255,7 +255,7 @@ func testAccCheckOAPIVirtualGatewayExists(n string, ig *oscgo.VirtualGateway) re
 				Filters: &oscgo.FiltersVirtualGateway{VirtualGatewayIds: &[]string{rs.Primary.ID}},
 			}).Execute()
 			if err != nil {
-				if strings.Contains(err.Error(), "RequestLimitExceeded:") {
+				if strings.Contains(err.Error(), "Request rate exceeded") {
 					return resource.RetryableError(err)
 				}
 				return resource.NonRetryableError(err)
```
Comment (outscale-mdr) [inline outscale/resource_outscale_virtual_gateway_test.go:261]: Why don't you use the refactored functions ?

File: outscale/resource_outscale_subnet.go [hunk 4]
```diff
@@ -135,14 +129,12 @@ func resourceOutscaleOAPISubNetDelete(d *schema.ResourceData, meta interface{})
...
 		_, _, err = conn.SubnetApi.DeleteSubnet(context.Background()).DeleteSubnetRequest(req).Execute()
 		if err != nil {
-			if strings.Contains(err.Error(),
-				"RequestLimitExceeded:") ||
-				strings.Contains(err.Error(), "Conflict") {
-				log.Printf("[DEBUG] Subnet waiting delete: (%s)",
-					err)
+			return utils.CheckThrolling(err)
+			if strings.Contains(err.Error(), "Conflict") {
+				log.Printf("[DEBUG] Subnet waiting delete: (%s)", err)
```
Comment (outscale-mdr) [inline outscale/resource_outscale_subnet.go:134]: LGTM warning

File: outscale/resource_outscale_snapshot.go [hunk 5]
```diff
@@ -283,30 +278,21 @@ func resourceOutscaleOAPISnapshotDelete(d *schema.ResourceData, meta interface{}
...
-
 			return nil
 		})
 		if err == nil {
 			return nil
 		}
 
-		ebsErr, ok := err.(awserr.Error)
+		ebsErr, _ := err.(awserr.Error)
 		if ebsErr != nil {
```
Comment (outscale-mdr) [inline outscale/resource_outscale_snapshot.go:290]: aws ?

File: outscale/resource_outscale_keypair.go [hunk 6]
```diff
@@ -102,7 +100,7 @@ func resourceOAPIKeyPairRead(d *schema.ResourceData, meta interface{}) error {
 		resp, _, err = conn.KeypairApi.ReadKeypairs(context.Background()).ReadKeypairsRequest(req).Execute()
 
 		if err != nil {
-			if strings.Contains(err.Error(), "RequestLimitExceeded:") {
+			if strings.Contains(err.Error(), "Request rate exceeded") {
 				return resource.RetryableError(err)
 			}
 			return resource.NonRetryableError(err)
```
Comment (outscale-mdr) [inline outscale/resource_outscale_keypair.go:106]: refactored functions ?

**Review by outscale-mdr — CHANGES_REQUESTED** [review]
Sometimes the `CheckThrolling` is not used

**Comment by outscale-mzi** [comment]
Hello, 
It seems that this issue is not fixed yet:
╷
│ Error: [DEBUG] Error reading network (503 Service Unavailable)
│ 
│   with outscale_net.net01[8],
│   on MultipleNets.tf line 8, in resource "outscale_net" "net01":
│    8: resource "outscale_net" "net01" {
│ 

Could you plz check again?

File: outscale/resource_outscale_vpn_connection.go [hunk 7]
```diff
@@ -229,11 +235,10 @@ func vpnConnectionRefreshFunc(conn *oscgo.APIClient, vpnConnectionID *string) re
 				VpnConnectionIds: &[]string{*vpnConnectionID},
 			},
 		}
-
 		resp, _, err := conn.VpnConnectionApi.ReadVpnConnections(context.Background()).ReadVpnConnectionsRequest(filter).Execute()
 		if err != nil {
 			switch {
-			case strings.Contains(fmt.Sprint(err), "RequestLimitExceeded:"):
+			case strings.Contains(fmt.Sprint(err), "Request rate exceeded"):
```
Comment (outscale-mdr) [inline outscale/resource_outscale_vpn_connection.go:241]: You could have put this in utils.go as a constant, no ?

File: utils/utils.go [hunk 8]
```diff
@@ -3,15 +3,27 @@ package utils
...
 	oscgo "github.com/outscale/osc-sdk-go/v2"
 	"github.com/spf13/cast"
 
 	"github.com/aws/aws-sdk-go/aws"
+	"github.com/hashicorp/terraform-plugin-sdk/helper/resource"
 )
 
 // PrintToJSON method helper to debug responses
+const (
+	ResourceNotFound string = "doesn't exist."
```
Comment (outscale-mdr) [inline utils/utils.go:20]: Specify that this is a APi error message with for example ResourceNotFoundErrorMessage

File: outscale/resource_outscale_route.go [hunk 9]
```diff
@@ -449,16 +437,13 @@ func findResourceOAPIRoute(conn *oscgo.APIClient, rtbid string, cidr string) (*o
 	var err error
 	err = resource.Retry(2*time.Minute, func() *resource.RetryError {
 		resp, _, err = conn.RouteTableApi.ReadRouteTables(context.Background()).ReadRouteTablesRequest(findOpts).Execute()
-
 		if err != nil {
-			if strings.Contains(fmt.Sprint(err), "InvalidParameterException") || strings.Contains(fmt.Sprint(err), "RequestLimitExceeded") {
+			if strings.Contains(fmt.Sprint(err), utils.InvalidState) {
```
Comment (outscale-mdr) [inline outscale/resource_outscale_route.go:441]: **InvalidState** is the same error as **InvalidParameterException** ?

File: outscale/resource_outscale_net_peering_test.go [hunk 10]
```diff
@@ -253,7 +247,7 @@ const testAccOAPIVpcPeeringConfig = `
 // 				})
 
 // 			if err != nil {
-// 				if strings.Contains(err.Error(), "RequestLimitExceeded:") {
+// 				if strings.Contains(err.Error(), "Request rate exceeded") {
 // 					return resource.RetryableError(err)
 // 				}
 // 				return resource.NonRetryableError(err)
```
Comment (outscale-mdr) [inline outscale/resource_outscale_net_peering_test.go:253]: remove theses lines ?

File: outscale/resource_outscale_keypair_test.go [hunk 11]
```diff
@@ -105,32 +105,21 @@ func testAccCheckOutscaleOAPIKeyPairDestroy(s *terraform.State) error {
...
 			return nil
 		}
-
-		// Verify the error is what we want
-		ec2err, ok := err.(awserr.Error)
-		if !ok {
-			return err
-		}
-		if ec2err.Code() != "InvalidOAPIKeyPair.NotFound" {
-			return err
```
Comment (outscale-mdr) [inline outscale/resource_outscale_keypair_test.go:130]: We change the behaviour ?

File: outscale/resource_outscale_client_gateway.go [hunk 12]
```diff
@@ -184,7 +199,7 @@ func clientGatewayRefreshFunc(conn *oscgo.APIClient, gatewayID *string) resource
 		resp, _, err := conn.ClientGatewayApi.ReadClientGateways(context.Background()).ReadClientGatewaysRequest(filter).Execute()
 		if err != nil || len(resp.GetClientGateways()) == 0 {
 			switch {
-			case strings.Contains(fmt.Sprint(err), "RequestLimitExceeded:"):
+			case strings.Contains(fmt.Sprint(err), "Request rate exceeded"):
```
Comment (outscale-mdr) [inline outscale/resource_outscale_client_gateway.go:202]: Rerfactor with constant from utils.go

**Review by outscale-mdr — CHANGES_REQUESTED** [review]
You cannot sign with opensouce at outscale.com

File: utils/utils.go [hunk 13]
```diff
@@ -77,3 +90,13 @@ func IsResponseEmptyOrMutiple(rLen int, resName string) error {
 	}
 	return nil
 }
+
+func CheckThrottling(err error) *resource.RetryError {
+	rand.Seed(time.Now().UnixNano())
+	if strings.Contains(err.Error(), Throttled) {
+		randTime := (rand.Float32()*(randMax-randMin+1) + randMin) * 1000
```
Comment (outscale-mdr) [inline utils/utils.go:97]: Why "+1" ?

You will get [RandMin, RandMax +1)

**Review by outscale-mdr — CHANGES_REQUESTED** [review]

File: outscale/resource_outscale_keypair_test.go (hunk 11 above)
Comment (outscale-mdr) [reply outscale/resource_outscale_keypair_test.go:130]: We do not check that ?

**Review by outscale-mdr — CHANGES_REQUESTED** [review]

File: outscale/resource_outscale_keypair_test.go (hunk 11 above)
Comment (outscale-toa) [reply outscale/resource_outscale_keypair_test.go:130]: Error was already returned
But you are right, I will add this condition before return
Thanks,

File: utils/utils.go (hunk 13 above)
Comment (outscale-toa) [reply utils/utils.go:97]: Do not get 2 times the same number

**Review by outscale-mdr — APPROVED** [review]

**Pull request merged**